libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+  if (curr_time + duration_us > UINT32_MAX) {
+    TIM_SetCompare1(TIM2, curr_time + duration_us - UINT32_MAX);
+  } else {
+    TIM_SetCompare1(TIM2, curr_time + duration_us);
+  }
+  s_active_timer_duration = duration_us;
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+static uint32_t prv_soft_timer_update_period(void) {
+  uint32_t counter = TIM_GetCounter(TIM2);
+  if (counter - s_last_update_count > 0) {
+    return counter - s_last_update_count;
+  } else {
+    return counter - s_last_update_count + UINT32_MAX;
+  }
+}
+
+static void prv_soft_timer_update(uint32_t time_since_update) {
+  // Default these values to max. If there is ever a newer timer it will replace them. In the event
+  // there are no timers left the defaults will result in the module being reset to a state where no
+  // timers exist.
+  s_last_update_count = TIM_GetCounter(TIM2);
+  SoftTimerID min_time_id = SOFT_TIMER_MAX_TIMERS;
+  uint32_t min_duration = UINT32_MAX;
+
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    // If the timer is in use and wasn't just added then update its duration.
+    if (s_soft_timer_array[i].inuse) {
+      if (time_since_update >= s_soft_timer_array[i].duration) {
+        // For each inuse timer see if the timer expired. If it has run its callback.
+        s_soft_timer_array[i].callback(i, s_soft_timer_array[i].context);
+        s_soft_timer_array[i].inuse = false;
+      } else {
+        // Otherwise, update the duration left on the timer.
+        s_soft_timer_array[i].duration -= time_since_update;
+        // Figure out the next timer.
+        if (min_duration >= s_soft_timer_array[i].duration) {
+          min_time_id = i;
+          min_duration = s_soft_timer_array[i].duration;
+        }
+      }
+    }
+  }
+  // Set the active timer id.
+  s_active_timer_id = min_time_id;
+}
+
+void soft_timer_init(void) {
+  // Start the PeiphClock
+  RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM2, ENABLE);
+  // Stop the timer if it was running and ensure only overflows trigger updates.
+  // Also remove the ARR preload.
+  TIM_Cmd(TIM2, DISABLE);
+  // TIM_InternalClockConfig(TIM2);
+
+  // Get the clock speed of the clocks to determine the SYSCLK speed which is what TIM2 uses.
+  RCC_ClocksTypeDef clock_speeds;
+  RCC_GetClocksFreq(&clock_speeds);
+
+  // Configure each clock tick to be 1 microsecond from 0 to duration.
+  TIM_TimeBaseInitTypeDef init_struct = { .TIM_Prescaler = clock_speeds.PCLK_Frequency / 1000000,
```

Comment:
Depending on how the clock manager is set up, I think you'll find that TIM2 defaults to 2x or 4x APB1. :P
I don't know what PCLK is defined as, but you might want verify this with a scope. If you make a periodic timer (I suggest starting with 1ms, then moving to 10us and hopefully 5us) that just toggles a GPIO, you should be able to measure the signal's period. You'll need to half it since it'll represent 2 callbacks, but it should give you a good idea of how accurate your timers are.